lib/preload.js

```diff
@@ -382,13 +382,56 @@ class Preloader extends EventEmitter {
 			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `statusCode` field (integer) and
+	 * optionally `result` and `error` fields. If `statusCode` is the number
+	 * zero, the command execution is considered successful, otherwise it is
+	 * assumed to have failed and the returned promise is rejected with an
+	 * error message including the message provided in the `error` field.
+	 *
+	 * @param {string} command
+	 * @param {object} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
 			this.stdout.once('data', (data) => {
-				resolve(JSON.parse(data).result);
+				let strData = data;
+				try {
+					strData = data.toString();
+				} catch (_e) {
+					// ignore
+				}
+				let response = {};
+				try {
+					response = JSON.parse(strData);
+				} catch (error) {
+					response.statusCode = 1;
+					response.error = error;
+				}
+				if (response.statusCode === 0) {
+					resolve(response.result);
+				} else {
+					const msg = [
+						`An error has occurred executing internal preload command '${command}':`,
+						cmd,
+					];
+					if (response.error) {
+						msg.push(
+							`Status code: ${response.statusCode}`,
+							`Error: ${response.error}`,
+						);
+					} else {
+						msg.push(`Response: ${strData}`);
```

Comment:
If `response.error` is set, it's a case of "graceful" error reporting from the Python side. If it is not set, something unexpected happened, so we just collect the raw `strdData`, hoping that it helps with debugging.